Add new "Going into production" doc

mechanomi opened the pull request:
## Summary of the changes / Why this is an improvement

at the moment, we have no clear set of instructions that help users know what they must change when going into production

## Checklist

 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed

mechanomi commented:
@joemoe @mfussenegger this PR is ready for review

you might find it easier to check out this branch and get the docs running locally to review

mfussenegger commented on docs/going-into-production.rst:
```diff
@@ -0,0 +1,175 @@
+
+.. _prod-auto-boot:
+
+Manual bootstrapping
+====================
+
+:ref:`Single host auto-bootstrapping <auto-bootstrapping>` is useful for
+development environments. However, auto-bootstrapping is typically too slow for
```
Feels to me a bit as if this is jumping in too quickly. Is there enough context established at this point so that readers know about the bootstrapping process ?

I'm also not sure about the argument that

> auto-bootstrapping is typically too slow

I think the speed of the bootstrapping process itself can be neglected. The primary reason is to ensure that in production you've the right configuration (multiple nodes that are eligible as master nodes, and that all are part of the  election process)

mechanomi replied:
okay expanded this section a lot. will commit a fixup in a min

mfussenegger commented on docs/going-into-production.rst:
```diff
@@ -0,0 +1,175 @@
+Paths
+-----
+
+In a production environment, you should change the default locations of the
+data and log folders to avoid accidental deletion.
```
Maybe add a note about disk sizing and disk IO?

joemoe replied:
Are there any typical locations you would chose? Can we give a recommendation?

mfussenegger commented on docs/going-into-production.rst:
```diff
@@ -0,0 +1,175 @@
+------------
+
+By default, CrateDB binds to the loopback address (i.e., `localhost`_). It
+listens on port 4200-4300 for HTTP traffic and port 4300-4400 for node-to-node
```
```suggestion
listens on port 4200-4299 for HTTP traffic and port 4300-4399 for node-to-node
```

PostgreSQL (`5432-5532`) could also be mentioned

mfussenegger commented on docs/going-into-production.rst:
```diff
@@ -0,0 +1,175 @@
+communication. Because CrateDB uses a port range, if one port is busy, it will
+automatically try the next port.
+
+When using :ref:`multiple hosts <multi_node_setup>`, nodes must bind to a
+non-loopback address.
+
+You can bind to a non-loopback address with the `network.host`_ setting.
+
+.. SEEALSO::
+
+    `Host settings`_
+
+.. CAUTION::
+
+      Never expose an unprotected CrateDB node to the public internet
```
Would it make sense to also mention here that bootstrap checks are enforced once CrateDB is bound to a non-loopback address?

infoverload replied:
I think yes.  I wasn't sure whether it was the case!

mechanomi replied:
done

mfussenegger commented on docs/scaling/multi-node-setup.rst:
```diff
@@ -4,162 +4,237 @@
+For development purposes, CrateDB can :ref:`auto-bootstrap
+<auto-bootstrapping>` a cluster when you run all nodes on the same host.
+Auto-bootstrapping requires zero configuration. However, in production
+environments, you must the bootstrapping process :ref:`manually
```
> you must the bootstrapping process

Is there a word missing here? You must configure the bootstrapping process, or something like that?

infoverload replied:
you must configure the bootstrapping process

infoverload commented on docs/going-into-production.rst:
```diff
@@ -0,0 +1,175 @@
+------------------
+
+CrateDB logs JVM garbage collection times using the built-in garbage collection
+logging of the JVM. You can configure this process the garbage collection
```
You can configure this process WITH the garbage collection

mfussenegger commented on docs/scaling/multi-node-setup.rst:
```diff
@@ -4,162 +4,237 @@
+
+    Each directory will function as `CRATE_HOME`_ for that node
+
+Because you want to run a multi-node cluster, you must configure the metadata
+gateway so that CrateDB knows how to recover its state. For a three-node cluster,
+set `gateway-expected-nodes`_ to ``3`` and set `gateway-recover-after-nodes`_
+to ``3``.
```
For the development auto-bootstrapping process this is not strictly necessary. It's more of a resiliency /safe-guard for production environments, to make sure a subset of the nodes don't recover a slightly outdated state without waiting that all nodes have been discovered.

mechanomi replied:
how about this:

> Configuring the metadata gateway is a safeguarding mechanism that is useful for production clusters. It is not strictly necessary when running in development. However, the Admin UI will issue warnings if the metadata gateway is not configured.

I added this step here because I don't want a reader's first experience with a CrateDB cluster to involve an Admin UI issuing warnings. that could alarm/concern some people

pushing a fixup

infoverload commented on docs/scaling/multi-node-setup.rst:
```diff
@@ -4,162 +4,237 @@
+.. WARNING::
+
+    If you start multiple nodes on different hosts with auto-bootstrapping
+    enabled, you cannot, at a later point, bring those nodes together for form
```
for form = to form

mfussenegger commented on docs/scaling/multi-node-setup.rst:
```diff
@@ -4,162 +4,237 @@
+~~~~~~~~~
 
-.. _master-node-election-4x:
+With CrateDB 4.x and above, you can configure a list of master-eligible nodes
```
The nodes here are not necessarily the `master-eligible` nodes, but this can include any kind of nodes.

infoverload commented on docs/scaling/multi-node-setup.rst:
```diff
@@ -4,162 +4,237 @@
+      2. Shut down all the nodes
+      3. Completely wipe each node by deleting the contents of the ``data``
+         directory under `CRATE_HOME`_
+      4. Follow in instructions in the next section (:ref:`manual bootstrapping
```
Follow in instructions = Follow the instructions

mfussenegger commented on docs/scaling/multi-node-setup.rst:
```diff
@@ -4,162 +4,237 @@
+    `discovery.zen.ping.unicast.hosts`_ setting instead of
+    `discovery.seed_hosts`_.
+
+CrateDB also supports node discovery via `DNS`_ as well as discovery via API
```
I'm wondering if this should be promoted a bit more?

mechanomi replied:
expanded this section a lot

joemoe commented on docs/going-into-production.rst:
```diff
@@ -0,0 +1,198 @@
+Switching to a manual bootstrapping configuration is the first step towards
+going into production. The rest of this how-to guide covers additional things
+you may want to consider.
+
```
this version is much better!

joemoe commented on docs/going-into-production.rst:
```diff
@@ -0,0 +1,198 @@
+
+.. _prod-config:
+
+Additional considerations
```
We are two headlines into the article and it already says "Additional considerations"?

mechanomi replied:
reworking this

joemoe commented on docs/going-into-production.rst:
```diff
@@ -0,0 +1,198 @@
+
+The `cluster.name`_ setting allows you to create multiple separate clusters. A
+node will refuse to join a cluster if the respective cluster names do not
+match.
```
Shouldn't we change the perspective? 
Most of the users want to create a single cluster and not multiple ones.
As this is a how-to guide once again I'd add some recommendations. What name can they use? What is a good name? What not?

joemoe commented on docs/going-into-production.rst:
```diff
@@ -0,0 +1,198 @@
+node names explicitly, you can specify a list of master-eligible nodes
+up-front.
+
+You can configure a custom node name using the `node.name`_ setting.
```
This anchor doesn't work or is there a separate pr?

mechanomi replied:
yeah. this will work once https://github.com/crate/crate/pull/10141 is merged

joemoe commented on docs/going-into-production.rst:
```diff
@@ -0,0 +1,198 @@
+Node name
+---------
+
+By default, CrateDB sets the node name for you. However, if you configure the
```
Can we say something about the fancy auto generated names?

joemoe commented on docs/going-into-production.rst:
```diff
@@ -0,0 +1,198 @@
+
+By default, CrateDB configures the JVM to dump out of memory exceptions to the
+file or directory specified by `CRATE_HEAP_DUMP_PATH`_. You must make sure
+there is enough disk space available for heap dumps at this location.
```
Do we really want to talk about heap dumps at this point?
To me this would be another - really cool - how-to: debugging CrateDB.

mechanomi replied:
that's a cool idea for a new doc!! created an issue here: https://github.com/crate/crate-howtos/issues/209

however, I want to leave this in. the goal here is not to get into how to debug crate. it is to alert syadmins to the fact that they need to consider disk space requirements for this directory

leaving this in

mechanomi commented:
@joemoe I have pushed some fixups that address your comments

- I have restructured the doc to use imperative titles
- I have added a lot more narrative
- I have added a lot more examples

inline comments have been responded to directly

I have not added an outro. we currently only do that for tutorials and blog posts where there is a strong narrative (i.e., we are guiding the reader through a step-by-step process). how-to guides are looser, and there is much more scope for the reader to pick and choose which sections they pay attention to. because of this, a useful "congratulations you did it" type ending isn't possible imo

let me know what you think!

mechanomi commented:
@infoverload @mfussenegger can I have your re-review, please?

infoverload commented on docs/going-into-production.rst:
```diff
@@ -0,0 +1,353 @@
+configuring a explicit node name using the `node.name`_ setting in your
+`configuration`_ file.
+
+Automatic nodes names are chosen the `sys.summits`_ table. This is useful in
```
missing a word in the sentence it seems

infoverload commented on docs/going-into-production.rst:
```diff
@@ -0,0 +1,353 @@
+`configuration`_ file.
+
+Automatic nodes names are chosen the `sys.summits`_ table. This is useful in
+development when you don't care too much about cluster architecture. However,
```
"when you don't care too much about cluster architecture" --> phrasing sounds a bit weird?

infoverload commented on docs/going-into-production.rst:
```diff
@@ -0,0 +1,353 @@
+    node.name: node-01-md
+
+Here, ``node`` is used as a base name with a sequence number of ``01``. Every
+node in the cluster should have a unique sequence number, irrespective of node
```
I feel like "irrespective" might be a weasel word but am not sure.

mechanomi replied:
switched to "independent"

infoverload commented on docs/going-into-production.rst:
```diff
@@ -0,0 +1,353 @@
+
+.. code-block:: yaml
+
+    node.name: node-01-m
```
should be changed to "node-03-m"

infoverload commented on docs/going-into-production.rst:
```diff
@@ -0,0 +1,353 @@
+The JVM uses a heap for memory allocations. For optimal performance, you must
+pay special attention to your :ref:`heap configuration <memory>`.
+
+By default, CrateDB configures the JVM to dump out of memory exceptions to the
```
out of memory = out-of-memory

mfussenegger commented on docs/going-into-production.rst:
```diff
@@ -0,0 +1,357 @@
+defaults to the installation directory). When you upgrade CrateDB, you will
+have to switch to a new installation directory.
+
+Instead of migrating data by hand each time, you should move the data
+directories off to a persistent location. You can do this using the
+`CRATE_HOME`_ environment variable and the `path settings`_ in your
+`configuration`_ file.
```
Should we maybe point out that the distribution packages include a service file that typically set up `CRATE_HOME` correctly, and that then it is only necessary to set the `path_settings`?

mechanomi replied:
done

joemoe commented on docs/going-into-production.rst:
```diff
@@ -0,0 +1,357 @@
+a good chance you have been using :ref:`single host auto-bootstrapping
+<auto-bootstrapping>`.
+
+For improved performance and resiliency, you should run production CrateDB
```
Should we include an `we recommend to use at least three nodes`?
I think we should mention this here, because that's what we recommend and there is no dedicated place for this information yet. Of course it is mentioned in the multi-node setup, but to get there you need to know you want to do this.

joemoe commented on docs/going-into-production.rst:
```diff
@@ -0,0 +1,357 @@
+--------------------------------
+
+The `cluster.name`_ setting allows you to override the default cluster name of
+``crate``. You should use this setting to give a logical name to your cluster.
```
This is maybe something for the configuration section, but do users need to follow a pattern?

mechanomi replied:
I've added a note suggesting a pattern directly under this para

joemoe commented on docs/going-into-production.rst:
```diff
@@ -0,0 +1,357 @@
+
+.. _prod-node-md:
+
+Multi-purpose nodes
```
Why is this a sub section of `Configuring logical node names`?
Doesn't make sense to me.

mechanomi replied:
because when configuring local node names per the system I have suggested, there are four types of nodes. I had to come up with descriptive names for each of the four types of node. the first node type I cover is the one with `master` and `data` set to `true`. that makes it a multi-purpose node

can you suggest a better description for this type of node? "master-eligible" doesn't cover it because you can have master-eligible nodes that don't have `data` set to true

to understand this section header you need to see it in context of the other three headers:

- Multi-purpose nodes
- Request handling and query execution nodes
- Cluster management nodes
- Request handling nodes

joemoe replied:
ok, somehow read it wrong.

joemoe commented on docs/going-into-production.rst:
```diff
@@ -0,0 +1,357 @@
+- ``example.com`` -- Your example domain name
+- ``internal`` -- Your internal private network
+- ``acme-prod`` -- The cluster name
+- ``node-01-md`` -- The node name
```
The DNS can't be configured here, this is given and defined somewhere else. 
To me this reads as if I have the option here.

The break down doesn't really provide any value, because this could be `anything`. Of course we can recommend how to do it, but then we should make it clear. As said, I'd even then doubt the value.

mechanomi replied:
I have clarified the point about setting this in DNS

I have also clarified the point that you should use the hostname to logically describe the node

as for the value of doing this: I added this stuff because in your previous review you requested that I provide suggestions re how to name stuff

joemoe commented on docs/going-into-production.rst:
```diff
@@ -0,0 +1,357 @@
+send debug logs to `STDERR`_ so that the container orchestrator handles the
+output.
+
+
```
Still don't like the end, I get your point, but maybe we can add something like further steps linking to administration,...

mechanomi replied:
I am sticking by my guns on this. I have reviewed our collection of how-to guides. I do not see any specific document or section that makes sense as a logical follow-on. at best, we would be guessing what the reader wants to do next

the tutorials are designed to follow-on from each other. i.e., we expect users to start at the beginning and work their way through

however, the how-to guides are more like the reference. they are designed for readers to dip in and dip out as needed. we do not expect people to read them sequentially

the how-to guides and the reference are roughly ordered by how advanced the topic is and how important/central the topic is. but that does not imply a particular "flow" through the docs

joemoe replied:
you should drop your guns

mechanomi commented:
pushed a fixup

some comments replied to inline

@joemoe, I am happy with the titles. they are imperative instructions (our house style) and they follow a logical order (i.e., they build on each other in order, and fit the order a reader would want to bring their attention to each issue). do you have a concrete suggestion for an alternative structure?

mfussenegger commented on docs/going-into-production.rst:
```diff
@@ -267,19 +276,38 @@ Then, you could configure your data paths like this:
+
+    If you have installed CrateDB using a system package for Debian, Ubuntu, or
+    Red Hat, the `CRATE_HOME`_ variable (as well as some other data paths) are
+    configured for by the `systemd`_ *service* file. View the source code
+    for `Debian`_, `Ubuntu`_, or `Red Hat`_ for the full details.
```
The repo is currently not public :/ 

But users can view the definition with `systemctl cat crate`

mechanomi replied:
ah! one sec

mechanomi replied:
pushed a fixup
